app/queries/get_grouped_content_and_links.rb

```diff
@@ -0,0 +1,116 @@
+module Queries
+  class GetGroupedContentAndLinks
+    PAGE_SIZE = 10
+    MAX_PAGE_SIZE = 1000
+    DEFAULT_CONTENT_ID = '00000000-0000-0000-0000-000000000000'.freeze
+
+    include ActiveModel::Validations
+
+    attr_accessor :last_seen_content_id, :page_size
+
+    validate :has_valid_uuid_for_pagination
+    validates_numericality_of :page_size, less_than_or_equal_to: MAX_PAGE_SIZE # FIXME: > 0
+
+    def initialize(last_seen_content_id: nil, page_size: nil)
+      @last_seen_content_id = last_seen_content_id || DEFAULT_CONTENT_ID
+      @page_size = page_size || PAGE_SIZE
+    end
+
+    # Returns an array of hashes where each hash contains a content_id, an
+    # array of content item rows, and an array of link rows.
+    def call
+      content_ids = query_content_ids_for_page(
+        last_seen_content_id: @last_seen_content_id,
+        page_size: @page_size
+      )
+
+      group_results(content_results(content_ids), link_set_results(content_ids))
+    end
+
+    def has_valid_uuid_for_pagination
+      unless @last_seen_content_id == DEFAULT_CONTENT_ID || UuidValidator.valid?(@last_seen_content_id)
+        errors.add(:last_seen_content_id)
+      end
+    end
+
+  private
+
+    def query_content_ids_for_page(last_seen_content_id:, page_size:)
+      ContentItem
+        .select(:content_id)
+        .group(:content_id)
+        .having("content_id > ?", [last_seen_content_id])
+        .limit(page_size)
+        .order(:content_id)
+        .map(&:content_id)
+    end
+
+    def group_results(content_results, link_set_results)
+      grouped_content_items = group_by_content_id(content_results)
+      grouped_links = group_by_content_id(link_set_results)
+
+      grouped_content_items.map do |content_id, content_items|
+        {
+          "content_id" => content_id,
+          "content_items" => content_items,
+          "links" => grouped_links[content_id] || []
+        }
+      end
+    end
+
+    def group_by_content_id(results)
+      results.group_by { |item| item["content_id"] }
+    end
+
+    def content_results(content_ids)
+      return [] if content_ids.empty?
+
+      query = <<-SQL
+        SELECT
+          ci.content_id as content_id,
```

Comment:
Can we use the names of the tables rather then the shortcut aliases, I think it is tricky to keep doing the conversion and helps with reading